### ISSUE TITLE
Updating phenotype data for several genes

### ISSUE THREAD

**Issue by sbslee** [sbslee/pypgx#36]
very interesting! I really admire how you already planned ahead for all of this situations.

I'm planning to add other functional impact of some other genes, such as: CACNA1S, RYR1, CFTR, CYP4F2, G6PD, IFNL3. I see the phenotype prediction page, as well as the categories in `PhenotypeMethod` of gene-table.tsv. I want to add this case to our discussion as well:

- CFTR: https://www.pharmgkb.org/chemical/PA165950341/guidelineAnnotation/PA166114461
In this guideline, the order of clinical action based on on genotyping will be:

G551D Het/Hom > F508del Homozygous > Some other variants Het/Hom. 

However, the meanings of those variant functions have not been defined in the list `FUNCTION_ORDER` yet. Do you think it's OK to add more entry for CFTR variant functionality score?

_Originally posted by @NTNguyen13 in https://github.com/sbslee/pypgx/issues/35#issuecomment-990869341_

**Comment by sbslee** [comment]
> For case such as CACNA1S and RYR1, I'm planning to add their `PhenotypeMethod` as `Diplotype`, and then add all of the possible combinations of 2 alleles to the `diplotype-table.csv` file (they must be sort using `sort_alleles` method, of course). Finally, I will add their phenotype based on PGKB annotation. Is this OK to do so?

**Comment by sbslee** [comment]
@NTNguyen13,

Thanks for asking before making PRs :) 

1. Could you tell me where you are planning to obtain diplotype-phenotype mapping for CACNA1S and RYR1? It doesn't look like such data exists in CPIC or PharmGKB. It'd be great if you can share some links first.
2. Same question as above for CYP4F2, G6PD, and IFNL3.
3. For CFTR I looked at the link you attached (thanks!) and it looks like we can use diplotype-phenotype mapping for phenotype prediction (ivacaftor recommended vs. not recommended). Is that what you have in your mind? If so, you will need to update the `diplotype-table.csv` file, not  `api.core.FUNCTION_ORDER`. The latter is used to give ordering between different enzymatic functions (e.g. `No Function` > `Normal Function`).

I think it's important that we have these discussions sooner than later because I hate to see you wasting your precious time.

**Comment by NTNguyen13** [comment]
For CACNA1S and RYR1, I think I can use the `Uncertain Susceptibility` and `Malignant Hyperthermia Susceptibility` phenotype from [here](https://www.pharmgkb.org/chemical/PA164749136/guidelineAnnotation/PA166180457). For example:

```
CACNA1S Diplotype: c.520C>T/Reference => Malignant Hyperthermia Susceptibility
CACNA1S Diplotype: Reference/Reference => Uncertain Susceptibility
RYR1 Diplotype: Reference/Reference => Uncertain Susceptibility
RYR1 Diplotype: c.742G>C/Reference => Malignant Hyperthermia Susceptibility
```

Similarly, IFNL3 phenotype can be determined using this [guide](https://www.pharmgkb.org/guidelineAnnotation/PA166110235):

```
IFNL3 Diplotype: rs12979860CT =>  unfavorable response
IFNL3 Diplotype: rs12979860TT =>  unfavorable response
IFNL3 Diplotype: rs12979860CC =>  favorable response
```

In CFTR case, what I'm concern about is when 1 individual has multiple alleles detected per chromosome (thanks again for your correction), such as F508del and G551D. The first issue is that, at this moment, all CFTR alleles are "Unknown Function" in `allele-table.csv`, so the function `sort_alleles` will not prioritize any of them. The second issue is the G551D allele should have the highest priority among all CFTR alleles(from my understanding from the [guideline](https://www.pharmgkb.org/chemical/PA165950341/guidelineAnnotation/PA166114461)), but its functionality meaning is not quite match with the current entries in `FUNCTION_ORDER` list.

To sum up, for CFTR, I think we should fill up its `Function` column in `allele-table.csv`, and the way we fill should be reflected in `FUNCTION_ORDER` list.

An example of the expected behavior: an individual with the diplotype of `F508del, G551D/F508del`, G511D should have higher functional priority than F508del, so the result of sorting and prioritizing process will be `G551D/F508del` => this individual can use ivacaftor as normal.

**Comment by NTNguyen13** [comment]
Hi, as I saw your work on IFNL3, CACNA1S and RYR1, I'm also curating data for G6PD genes according to this [guideline](https://www.pharmgkb.org/chemical/PA10176/guidelineAnnotation/PA166119846). At first I tried to do it with `PhenotypeMethod` as `Diplotype`, but 186 alleles will yield 17205 combinations, which make the `diplotype-table.csv` quite large. Additionally, this variant is on chrX, which leads to a discrepancy in the guideline for 2 genders.

I think that I can assign the activity score using this strategy based on Class I, Class II, Class III and Class IV:

- AS = 0 for Class I, Class II, Class III
- AS = 1 for Class IV

One male will always be assumed to carry the B alleles (Class IV), so we will only deal with 3 cases:

- 2 Class IV: AS equals to 2 => normal
- 1 Class I-III: AS equals to 1 => variable/unknown
- 2 Class I-III: AS equals to 0 => Deficient/Deficient with CNSHA

What do you think about my strategy? I can make the PR if you need my help.

**Comment by sbslee** [comment]
@NTNguyen13,

Thanks for the links and additional information!

1. I updated phenotype data for CACNA1S, IFNL3, and RYR1. Please let me know if you have any questions/concerns about this update.
2. I still couldn't find phenotype data for CYP4F2 and G6PD. If you can share links containing a phenotype table for these genes, that'd be great.
3. For CFTR I agree that we'd need to update both `allele-table.csv` and `FUNCTION_ORDER`. And that's exactly what I did for CACNA1S, IFNL3, and RYR1. However, I'm not so sure about giving more priority to G551D than F508del. In the example you gave, if an individual is homozygous for F508del ivacaftor is not recommended because "ivacaftor does not potentiate CFTR function in F508del-CFTR homozygotes (probably due to inadequate CFTR expression at the cell membrane)" according to the CPIC [paper](https://files.cpicpgx.org/data/guideline/publication/ivacaftor/2014/24598717.pdf). Also, CFTR variants are pretty rare so I'm not sure how likely we will find multiple variants in one haplotype (chromosome). If such haplotype exists (e.g. a haplotype with both G551D and F508del), then I think it should be reported to the world so that it can be defined as a novel haplotype. Were you asking these questions as hypothetical? Or do you actually have a sample with the `F508del,G551D/F508del` genotype?

**Comment by NTNguyen13** [comment]
Yes, my question is purely hypothetical, I didn't observe it among 1000 samples of our project.

**Comment by sbslee** [comment]
@NTNguyen13, thanks for the prompt answer. If that's the case, then I don't think we need to worry about the multiple-alleles-in-one-haplotype problem for CFTR right now :) And because of this, we don't need to update `allele-table.csv` for CFTR. That being said, do you want to make a PR updating phenotype data for CFTR? You can look at CACNA1S, IFNL3, and RYR1 to see which files need to be updated. But no pressure whatsoever. Just let me know if you want to give it a shot. Otherwise, I can take care of it pretty easily.

In the case of G6PD, I will have to review it more carefully because, as you said, it's on the X chromosome. Therefore, updating phenotype data for G6PD requires not just addition of some new data but also potentially new structuring of the code to, for example, allow specification of sex, etc. for phenotype prediction. I will get back to you on this soon.

**Comment by NTNguyen13** [comment]
Hi, I made the PR for CFTR. I tried to follow your format, but there may be something that I didn't notice.

P/s: I just notice on the PR that git diff the whole `pypgx/api/data/diplotype-table.csv`. It's quite weird, on my vscode it only shows that I concatenated CFTR at the end of the file. Do you need me to create another PR?

**Comment by NTNguyen13** [comment]
I noticed that the original `pypgx/api/data/diplotype-table.csv` does not have the newline character at the end of file, while my file in #37 have it, so git diffed the whole file. I removed it and created a new PR #38
